### PR TITLE
Implement aggregate constraints

### PR DESCRIPTION
**Pull request by Pyrofab**
This PR reworks constraints to bring support for arrays and collections. I chose to refer to those types as "Aggregate" and to everything else as "Scalar", as reflected in the new `ConstraintsBuilder` and `ConfigValueBuilder` specializations. Elements of an aggregate are called "Components", like the components of an array.

Main features :
- Every scalar constraint can now be applied to components of arrays and collections.
- The length constraints have been refactored to work on aggregate types on top of `CharSequence`
    - Bonus : shorter names !

Cons : 
- The generics get a bit out of hand sometimes
- The syntax for array component constraint annotation is really obscure
```java
@Setting.Constrain.BiggerThan(2) int private @Setting.Constrain.MinLength(1) [] a = {3};
```

### CODE REVIEW

File: src/test/java/me/zeroeightsix/fiber/builder/constraint/ConstraintsBuilderTest.java [hunk 1]
```diff
@@ -38,4 +40,22 @@ public void testNumericalConstraints() {
...
+                .finishComponent()
+                .maxLength(3)
+                .finish()
+                .build();
+
+        assertTrue(config.setValue(new Integer[0]));
+        assertTrue(config.setValue(new Integer[]{4, 5, 6}));
+        assertFalse(config.setValue(new Integer[]{1, 2}));
+        assertFalse(config.setValue(new Integer[]{5, 6, 7, 8}));
+        assertFalse(config.setValue(new Integer[]{9, 10, 11}));
```
Comment (zeroeightysix) [inline src/test/java/me/zeroeightsix/fiber/builder/constraint/ConstraintsBuilderTest.java]: Can you also add a test for aggregate/component constraints on collections?

**Comment by zeroeightysix** [comment]
This looks pretty good so far. I'm not too keen on the generic clusterfuck, but it's somewhat unavoidable and you've done a great job of hiding it from the API users.

Now, the one thing you're going to hate me for: we're not 'monitoring' additions to collections, while we _could_ be doing so.

What I had in mind is, when you create a `ConfigValue` of some generic `Collection` - Fiber actually returns a `ConstrainedCollection` (see below), with the original collection as backing collection.

A `ConstrainedCollection` would be:
```java
class ConstrainedCollection<T> implements Collection<T> {
    @Nonnull
    private final Collection<T> backingCollection;
    @Nonnull
    private final Predicate<T> constraint;

    public ConstrainedCollection(@Nonnull Collection<T> backingCollection, @Nullable Predicate<T> constraint) {
        this.backingCollection = backingCollection;
        this.constraint = (constraint == null) ? (ignored) -> true : constraint;
    }

    @Override
    public boolean add(T t) {
        if (!constraint.test(t)) return false;
        return backingCollection.add(t);
    }

    @Override
    public boolean addAll(Collection<? extends T> collection) {
        return backingCollection.addAll(collection.stream()
                .filter(element -> !constraint.test(element))
                .collect(Collectors.toList()));
    }

    // All other methods are simply passed to the backing collection:
    @Override
    public boolean remove(Object o) {
        return backingCollection.remove(o);
    }
    //etc...
}
```

This way we can replicate the **expected** behaviour of also constraining the incoming values. Of course, returning this kind of collection shouldn't be mandatory (but imo, should be default).

Of course I could be overlooking a million problems here. What are your thoughts?

File: src/main/java/me/zeroeightsix/fiber/annotation/AnnotatedSettings.java [hunk 2]
```diff
@@ -1,16 +1,18 @@
 package me.zeroeightsix.fiber.annotation;
 
+import blue.endless.jankson.magic.TypeMagic;
```
[the remark this answers is not in the document]
Comment (Pyrofab) [reply src/main/java/me/zeroeightsix/fiber/annotation/AnnotatedSettings.java]: What should we do then, duplicate the method ?

[the remark this answers is not in the document]
Comment (zeroeightysix) [reply src/main/java/me/zeroeightsix/fiber/annotation/AnnotatedSettings.java]: Just looked at the method you're using - it's a little too long to duplicate. Let's just leave this in, and solve the problem if jankson is ever separated.

**Comment by Pyrofab** [comment]
This is a pretty reasonable suggestion. I think we should take this a step further and allow arbitrary lenses when building a `ConfigValue` - lambdas that convert the actual value to a constrained form.

**Comment by zeroeightysix** [comment]
Thanks!

**Pull request merged**